[night-issues: Cycle Verify] Maven: system PSH2MultiuserLockHarnessTest

## Cycle Verify residual (next-cycle lead)

This is **not a qa task**. Do not assign human QA. Unassigned p1 IMPLEMENT residual for the next night-issue-prs Work cycle.

### integration_ref
- Branch: `cluster/night-issue-20260901-developer-catalog-persist`
- SHA: `72f10c55085cfd9a64041f5bbbd7269726f8a72e`
- Cluster PR: https://github.com/intersoftdatalabs-in/percussioncms/pull/4104

### Command
```
cd system
../mvnw.cmd clean install
```
(no `-DskipTests`; JDK 21 via JAVA_HOME; worktree `night-issue-prs`)

### First failure
```
[ERROR] Failures:
[ERROR]   PSH2MultiuserLockHarnessTest.concurrentSameItemCheckout_singleWinner:262 Exactly one exclusive checkout must succeed ==> expected: <1> but was: <2>
[ERROR] Tests run: 2702, Failures: 1, Errors: 0, Skipped: 247
[INFO] BUILD FAILURE
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.4:test (default-test) on project perc-system: There are test failures.
```

Test: `com.percussion.services.datasource.PSH2MultiuserLockHarnessTest.concurrentSameItemCheckout_singleWinner` (H2 multiuser lock harness #548 T009/T069/T070). Two exclusive checkouts succeeded under concurrent load; the harness requires a single winner.

This is **not** in tonight's Developer catalog persist diff (views/searches/display formats). It still failed on the integration tip clean install.

### This-run PRs
- Cluster: https://github.com/intersoftdatalabs-in/percussioncms/pull/4104 (absorbed #4088 #4094 #4099 #4100 #4102 #4103)
- https://github.com/intersoftdatalabs-in/percussioncms/pull/4088
- https://github.com/intersoftdatalabs-in/percussioncms/pull/4099
- https://github.com/intersoftdatalabs-in/percussioncms/pull/4100
- https://github.com/intersoftdatalabs-in/percussioncms/pull/4102
- https://github.com/intersoftdatalabs-in/percussioncms/pull/4103

### Other modules this cycle (green)
- rest: BUILD SUCCESS Tests run: 963 Failures: 0
- projects/sitemanage: BUILD SUCCESS Tests run: 2123 Failures: 0 Skipped: 125
- WebUI: BUILD SUCCESS Surefire 63 / Vitest 3675
- modules/perc-qa-automation: BUILD SUCCESS
- Playwright surface (H2 QA): 28 passed (golden + login + developer catalog/community/display-format/view/search)

### Acceptance
- `cd system && ../mvnw.cmd clean install` (or `../mvnw` on Unix) is **BUILD SUCCESS**
- `PSH2MultiuserLockHarnessTest` (especially `concurrentSameItemCheckout_singleWinner`) is green; if the race is real, tighten exclusive checkout so only one winner; if the assertion is racy, fix the harness so it is deterministic on Windows + Unix
- not a qa task

> Co-Authored by Grok Build 1.0.5 using grok-4.6 with agent night-issue-prs-cycle-verify.

## Agent progress (night-issue-prs)

| Slice | Issue | Status | PR | Notes | Updated |
|-------|-------|--------|----|-------|---------|
| H2 multiuser lock harness single-winner checkout | #4105 | pr_opened | #4106 | CAS checkout + hold-until-attempted; system clean install BUILD SUCCESS | 2026-09-01 |

--- a/docs/ai-generated/code-reviews/4105-h2-multiuser-lock-single-winner-erlang.md
+++ b/docs/ai-generated/code-reviews/4105-h2-multiuser-lock-single-winner-erlang.md
@@ -1,0 +1,49 @@
+# Erlang review: #4105 H2 multiuser lock harness single-winner checkout
+
+**Branch:** `fix/issue-4105-h2-multiuser-lock-single-winner`  
+**Base:** `origin/main`  
+**Scope:** uncommitted `PSH2MultiuserLockHarnessTest`  
+**Recommendation:** approve  
+**Gate:** May commit/push: yes  
+**Memory patterns hit:** tests.behavioral (CAS rowcount + hold-until-attempted latch); paths.hardcoded-sep false-positive (H2 JDBC URL `/`)
+
+## Summary
+
+Test-only harness fix for exclusive content checkout on file H2 (`FILE_LOCK=NO`). `SELECT FOR UPDATE` plus an unguarded `UPDATE` allowed two concurrent winners (and/or a serial handoff after check-in inflated the count). Checkout is now a compare-and-set `UPDATE … WHERE CHECKOUTUSER IS NULL` (rowcount == 1); the same-item test holds the winner's checkout until all editors have attempted. Serial handoff after check-in remains allowed and is covered separately.
+
+## Scope
+
+- Base: `origin/main` (`4f8c8dd957`)
+- Head: `fix/issue-4105-h2-multiuser-lock-single-winner` (uncommitted)
+- Files: 1 production-adjacent test + this review
+- Prior report: none
+- Memory patterns hit: tests.behavioral; paths JDBC URL `/` is not filesystem join
+
+## Recommendation
+
+approve
+
+## Gate
+
+- Blocking bugs: 0
+- May commit/push: yes
+
+## Issues
+
+None.
+
+## Cross-platform path checklist
+
+Applied. Existing JDBC URL still uses `Path` + `toAbsolutePath()` and replaces `\` with `/` only for the H2 `file:` URL (URL path, not OS filesystem join). `@TempDir` is portable. No new hardcoded separators, Unix-only roots, or line-ending assertions.
+
+## Tests
+
+- `PSH2MultiuserLockHarnessTest`: 7 run, 0 fail (focused `mvnw -Dtest=…`)
+- `cd system && ../mvnw.cmd clean install`: BUILD SUCCESS; Tests run: 2692, Failures: 0, Errors: 0, Skipped: 247
+- Change-class companions: same-class CAS contention + serial-handoff tests; no production API / Spring context / product-docs / Playwright
+
+## Notes
+
+- C2 reverse-deps: N/A (test-only, no `final`/signature change)
+- C5 Playwright: N/A (no UI)
+- Product documentation: N/A (test-only harness)

--- a/system/src/test/java/com/percussion/services/datasource/PSH2MultiuserLockHarnessTest.java
+++ b/system/src/test/java/com/percussion/services/datasource/PSH2MultiuserLockHarnessTest.java
@@ -51,8 +51,8 @@ import org.junit.jupiter.api.io.TempDir;
  * without a full CMS Spring container:
  *
  * <ul>
- *   <li>Exclusive checkout via {@code SELECT … FOR UPDATE} then update of a checkout-user column
- *       (CONTENTSTATUS-style)
+ *   <li>Exclusive checkout via {@code SELECT … FOR UPDATE} then a compare-and-set update of a
+ *       checkout-user column ({@code WHERE CHECKOUTUSER IS NULL} / rowcount == 1, CONTENTSTATUS-style)
  *   <li>Design-object locks via row-level exclusive claim (PSObjectLock-style)
  *   <li>≥10 concurrent "editors" on distinct connections (FR-003 / SC-005 floor)
  *   <li>Same-item exclusive vs distinct-item parallel (T070)
@@ -225,6 +225,9 @@ public class PSH2MultiuserLockHarnessTest {
     final int itemId = 15;
     ExecutorService pool = Executors.newFixedThreadPool(EDITOR_COUNT);
     CountDownLatch start = new CountDownLatch(1);
+    // Hold the winner's checkout until every editor has attempted, so a legitimate serial
+    // handoff after check-in cannot inflate the exclusive-winner count (#4105).
+    CountDownLatch attemptsDone = new CountDownLatch(EDITOR_COUNT);
     AtomicInteger winners = new AtomicInteger();
     AtomicInteger losers = new AtomicInteger();
     List<Future<?>> futures = new ArrayList<>();
@@ -234,11 +237,22 @@ public class PSH2MultiuserLockHarnessTest {
       futures.add(
           pool.submit(
               () -> {
+                boolean won = false;
                 try {
-                  start.await(30, TimeUnit.SECONDS);
-                  if (checkoutItem(itemId, "editor-" + editor)) {
-                    // hold lock briefly then edit
-                    Thread.sleep(50);
+                  if (!start.await(30, TimeUnit.SECONDS)) {
+                    throw new IllegalStateException("start latch timeout");
+                  }
+                  won = checkoutItem(itemId, "editor-" + editor);
+                } catch (Exception ex) {
+                  throw new RuntimeException(ex);
+                } finally {
+                  attemptsDone.countDown();
+                }
+                try {
+                  if (won) {
+                    if (!attemptsDone.await(30, TimeUnit.SECONDS)) {
+                      throw new IllegalStateException("checkout-attempt latch timeout");
+                    }
                     updateBody(itemId, "editor-" + editor, "winner-body-" + editor);
                     checkinItem(itemId, "editor-" + editor);
                     winners.incrementAndGet();
@@ -272,6 +286,80 @@ public class PSH2MultiuserLockHarnessTest {
         assertNull(rs.getString(1));
         assertEquals(1, rs.getInt(2), "single edit applied");
         assertTrue(rs.getString(3).startsWith("winner-body-"));
+      }
+      c.commit();
+    }
+  }
+
+  @Test
+  @DisplayName("After exclusive check-in, a different editor may check out (serial handoff)")
+  void checkoutAfterCheckin_nextEditorWins() throws Exception {
+    final int itemId = 16;
+    assertTrue(checkoutItem(itemId, "editor-0"));
+    updateBody(itemId, "editor-0", "first-edit");
+    checkinItem(itemId, "editor-0");
+    assertTrue(checkoutItem(itemId, "editor-1"), "check-in must release exclusive checkout");
+    updateBody(itemId, "editor-1", "second-edit");
+    checkinItem(itemId, "editor-1");
+
+    try (Connection c = connect();
+        PreparedStatement ps =
+            c.prepareStatement(
+                "SELECT CHECKOUTUSER, VERSION, BODY FROM CONTENT_ITEM WHERE CONTENTID = ?")) {
+      ps.setInt(1, itemId);
+      try (ResultSet rs = ps.executeQuery()) {
+        assertTrue(rs.next());
+        assertNull(rs.getString(1));
+        assertEquals(2, rs.getInt(2), "two serial edits");
+        assertEquals("second-edit", rs.getString(3));
+      }
+      c.commit();
+    }
+  }
+
+  @Test
+  @DisplayName("Compare-and-set checkout UPDATE yields a single rowcount winner under contention")
+  void atomicCheckoutClaim_rowcountSingleWinner() throws Exception {
+    final int itemId = 17;
+    ExecutorService pool = Executors.newFixedThreadPool(EDITOR_COUNT);
+    CountDownLatch start = new CountDownLatch(1);
+    AtomicInteger claims = new AtomicInteger();
+    List<Future<?>> futures = new ArrayList<>();
+
+    for (int e = 0; e < EDITOR_COUNT; e++) {
+      final String user = "cas-" + e;
+      futures.add(
+          pool.submit(
+              () -> {
+                try {
+                  if (!start.await(30, TimeUnit.SECONDS)) {
+                    throw new IllegalStateException("start latch timeout");
+                  }
+                  if (claimFreeCheckout(itemId, user) == 1) {
+                    claims.incrementAndGet();
+                  }
+                } catch (Exception ex) {
+                  throw new RuntimeException(ex);
+                }
+              }));
+    }
+
+    start.countDown();
+    for (Future<?> f : futures) {
+      f.get(60, TimeUnit.SECONDS);
+    }
+    pool.shutdown();
+    assertTrue(pool.awaitTermination(30, TimeUnit.SECONDS));
+
+    assertEquals(1, claims.get(), "exactly one CAS checkout claim");
+    try (Connection c = connect();
+        PreparedStatement ps =
+            c.prepareStatement("SELECT CHECKOUTUSER FROM CONTENT_ITEM WHERE CONTENTID = ?")) {
+      ps.setInt(1, itemId);
+      try (ResultSet rs = ps.executeQuery()) {
+        assertTrue(rs.next());
+        String holder = rs.getString(1);
+        assertTrue(holder != null && holder.startsWith("cas-"), holder);
       }
       c.commit();
     }
@@ -375,6 +463,9 @@ public class PSH2MultiuserLockHarnessTest {
   /**
    * Exclusive checkout: lock row, succeed only if free or already held by same user.
    *
+   * <p>H2 {@code SELECT … FOR UPDATE} is not a reliable exclusive barrier across connections in
+   * this FILE_LOCK=NO harness, so the claim is the compare-and-set {@code UPDATE} (rowcount == 1).
+   *
    * @return true if this editor holds the checkout after the call
    */
   private boolean checkoutItem(int contentId, String user) throws SQLException {
@@ -398,17 +489,46 @@ public class PSH2MultiuserLockHarnessTest {
           }
         }
       }
-      try (PreparedStatement upd =
-          c.prepareStatement("UPDATE CONTENT_ITEM SET CHECKOUTUSER = ? WHERE CONTENTID = ?")) {
-        upd.setString(1, user);
-        upd.setInt(2, contentId);
-        upd.executeUpdate();
+      int n = claimFreeCheckout(c, contentId, user);
+      if (n != 1) {
+        LAST_SQL_ERROR.set("claim lost (rowcount=" + n + ")");
+        c.rollback();
+        return false;
       }
       c.commit();
       return true;
     } catch (SQLException e) {
       LAST_SQL_ERROR.set(e.getSQLState() + " " + e.getMessage());
       return false;
+    }
+  }
+
+  /**
+   * Single-winner checkout claim: update only a free row (or a re-checkout by the same user).
+   *
+   * @return updated row count (1 = this connection won the claim)
+   */
+  private int claimFreeCheckout(int contentId, String user) throws SQLException {
+    try (Connection c = connect()) {
+      int n = claimFreeCheckout(c, contentId, user);
+      if (n == 1) {
+        c.commit();
+      } else {
+        c.rollback();
+      }
+      return n;
+    }
+  }
+
+  private int claimFreeCheckout(Connection c, int contentId, String user) throws SQLException {
+    try (PreparedStatement upd =
+        c.prepareStatement(
+            "UPDATE CONTENT_ITEM SET CHECKOUTUSER = ? WHERE CONTENTID = ? AND (CHECKOUTUSER IS NULL"
+                + " OR CHECKOUTUSER = '' OR CHECKOUTUSER = ?)")) {
+      upd.setString(1, user);
+      upd.setInt(2, contentId);
+      upd.setString(3, user);
+      return upd.executeUpdate();
     }
   }
 

